presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util.maps;
+
+import com.google.common.base.Equivalence;
+
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class IdentityLinkedHashMap<K, V>
+        implements Map<K, V>
+{
+    private final Map<Equivalence.Wrapper<K>, V> delegate = new LinkedHashMap<>();
+    private final Equivalence<Object> equivalence = Equivalence.identity();
+
+    public IdentityLinkedHashMap()
+    {
+    }
+
+    public IdentityLinkedHashMap(IdentityLinkedHashMap<K, V> map)
+    {
+        putAll(map);
+    }
+
+    @Override
+    public int size()
+    {
+        return delegate.size();
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(Object key)
+    {
+        return delegate.containsKey(equivalence.wrap(key));
+    }
+
+    @Override
+    public boolean containsValue(Object value)
+    {
+        return delegate.containsValue(value);
+    }
+
+    @Override
+    public V get(Object key)
+    {
+        return delegate.get(equivalence.wrap(key));
+    }
+
+    @Override
+    public V put(K key, V value)
+    {
+        return delegate.put(equivalence.wrap(key), value);
+    }
+
+    @Override
+    public V remove(Object key)
+    {
+        return delegate.remove(equivalence.wrap(key));
+    }
+
+    @Override
+    public void putAll(Map<? extends K, ? extends V> map)
+    {
+        map.entrySet().forEach(e -> {
+            delegate.put(equivalence.wrap(e.getKey()), e.getValue());
+        });
+    }
+
+    @Override
+    public void clear()
+    {
+        delegate.clear();
+    }
+
+    @Override
+    public Set<K> keySet()
+    {
+        return new AbstractSet<K>()
+        {
+            @Override
+            public Iterator<K> iterator()
+            {
+                return delegate.keySet().stream().map(Equivalence.Wrapper::get).iterator();
+            }
+
+            @Override
+            public int size()
+            {
+                return delegate.size();
+            }
+        };
+    }
+
+    @Override
+    public Collection<V> values()
+    {
+        return delegate.values();
+    }
+
+    @Override
+    public Set<Entry<K, V>> entrySet()
```

Comment:
The output of this method doesn't satisfy the required contract:

```
* Returns a {@link Set} view of the keys contained in this map.
* The set is backed by the map, so changes to the map are
* reflected in the set, and vice-versa.
```

It would be acceptable to make all other methods throw an exception, but having them silently succeed and not do the right thing is a recipe for future subtle bugs.